### DISCUSSION TITLE
keymap binds

### DISCUSSION THREAD

**Discussion by ConanEdogawa1** [LazyVim/LazyVim#287] [General]
I modified the key to fail after using insert mode in a new file. Keys work without editing, and keymappings fail after modifying files
<img width="635" alt="WeChata8944699dac461afe0339db57c3b30eb" src="https://user-images.githubusercontent.com/49578735/219611419-71d4e226-de95-4a59-a0a9-c135a40c955e.png">

In this configuration, u should be moved up, but in the newly edited file it will become undone

**Comment by folke** [comment]
Where did you add this? Please provide some context. The file should be lua/config/keymaps.lua

**Comment by ConanEdogawa1** [comment]
Added .lua lua/config/keymap

**Comment by ConanEdogawa1** [comment]
<img width="1413" alt="WeChat1d21de566000d6c5e4a62ed92d6e646c" src="https://user-images.githubusercontent.com/49578735/219620702-76d76f35-b0a2-40df-bc1c-97f7c9499eab.png">

**Comment by dpetka2001** [comment]
I think op wants to remap a key that has already been mapped. But the default behavior of `vim.keymap.set` is `remap=false`. So maybe he should pass `remap=true` to the mapping he wants to achieve?

**Reply by ConanEdogawa1** [reply]
I'm new to NVIM and don't know much

**Reply by dpetka2001** [reply]
Just try the mapping you want to do but also pass `remap=true` in the `opts`.

**Reply by ConanEdogawa1** [reply]
<img width="829" alt="WeChatc72d5f4413baa7903e7819eaaabd0429" src="https://user-images.githubusercontent.com/49578735/219621827-32902e31-d93e-4069-9044-e00d60d74e9c.png">

Is that so！！

**Reply by dpetka2001** [reply]
If that is not what you meant, please be more specific about what you want to achieve.

**Comment by dpetka2001** [comment]
Do you want to use your `u` key to move up and your `l` key to undo changes?

**Reply by ConanEdogawa1** [reply]
yep

**Reply by dpetka2001** [reply]
Try the following
```lua
map("n", "u", "<Up>")
map("n", "l", "<Undo>")
```

**Reply by ConanEdogawa1** [reply]
Thanks, I'll try

**Reply by dpetka2001** [reply]
And whatever `opts` you want to pass obviously. No need for the `remap` thing I mentioned before. Please disregard that.

**Comment by folke** [comment]
You need to add expr=true, sorry hadn't seen it before 

You also shouldn't use that special map function. Just use vim.keymap.set

**Reply by dpetka2001** [reply]
Yes I also have `map=vim.keymap.set`, so I guess I should have mentioned it before.

**Reply by ConanEdogawa1** [reply]
Sorry, sorry. I am using a colemak keyboard to refer to someone else's configuration <https://github.com/theniceboy/nvim/init.vim>

**Reply by dpetka2001** [reply]
@folke Could I ask why `expr=true` is needed? I tried without it as stated in the post above and the mappings worked.

**Reply by ConanEdogawa1** [reply]
Could it be that the plugin shortcut key conflict is caused, I still can't use your method

**Reply by dpetka2001** [reply]
I tried the keymaps above and it worked for me. Are you sure you did the same?
```lua
vim.keymap.set("n", "u", "<Up>")
vim.keymap.set("n", "l", "<Undo>")
```

**Reply by ConanEdogawa1** [reply]
-- Keymaps are automatically loaded on the VeryLazy event
-- Default keymaps that are always set: https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/config/keymaps.lua
-- Add any additional keymaps here

-- Coleman Keymaps Config
local Util = require("lazyvim.util")

local function map(mode, lhs, rhs, opts)
  local keys = require("lazy.core.handler").handlers.keys
  ---@cast keys LazyKeysHandler
  -- do not create the keymap if a lazy keys handler exists
  if not keys.active[keys.parse({ lhs, mode = mode }).id] then
    opts = opts or {}
    opts.silent = opts.silent ~= false
    vim.keymap.set(mode, lhs, rhs, opts)
  end
end

-- local opts = { noremap = true, silent = true, remap = true }
-- vim.keymap.set("n", "l", "u", opts)
-- vim.keymap.set("n", "u", "k", opts)
-- Move
map("n", "n", "h", { silent = true })
map("n", "e", "j", { silent = true })
map("n", "u", "k", { silent = true, expr = true, remap = true })
-- map("n", "u", "<Up>",expr=true)
-- map("n", "l", "<Undo>")
map("n", "i", "l", { silent = true })
map("n", "gu", "gk", { silent = true })
map("n", "ge", "gj", { silent = true })
map("n", "\v", "v$h", { silent = true })

-- visual mode
map("v", "n", "h", { silent = true })
map("v", "e", "j", { silent = true })
map("v", "u", "k", { silent = true })
map("v", "i", "l", { silent = true })

-- set h (same as n, cursor left) to 'end of word'
map("n", "h", "e", { silent = true })
-- Basic Mappings
map("n", "l", "u", { silent = true, expr = true, remap = true })
map("n", "k", "i", { silent = true })
map("n", "K", "I", { silent = true })
map("n", "Y", '"+y', { silent = true })
map("n", ";", ":", { silent = true })

-- Searching
map("n", "-", "N", { silent = true })
map("n", "+", "n", { silent = true })

-- " Indentation
map("n", "<", "<<", { silent = true })
map("n", ">", ">>", { silent = true })

-- U/E keys for 5 times u/e (faster navigation)
map("n", "U", "5k", { silent = true })
map("n", "E", "5j", { silent = true })

-- N key: go to the start of the line
-- I key: go to the end of the lines
map("n", "N", "0", { silent = true })
map("n", "I", "$", { silent = true })

-- Faster in-line navigation
map("n", "W", "5w", { silent = true })
map("n", "B", "5b", { silent = true })

-- set h (same as n, cursor left) to 'end of word'
map("n", "h", "e", { silent = true })

-- Ctrl + U or E will move up/down the view port without moving the cursor
map("n", "<C-U>", "5<C-y>", { silent = true })
map("n", "<C-E>", "5<C-e>", { silent = true })

-- Move to window using the <ctrl> nuei keys
map("n", "<C-n>", "<C-w>h", { desc = "Go to left window" })
map("n", "<C-e>", "<C-w>j", { desc = "Go to lower window" })
map("n", "<C-u>", "<C-w>k", { desc = "Go to upper window" })
map("n", "<C-i>", "<C-w>l", { desc = "Go to right window" })

**Reply by dpetka2001** [reply]
The shortcut for `Lazy` is `<leader>l` so there shouldn't be any conflict. Just restart your Neovim and open a file and check if the mappings work.

**Reply by dpetka2001** [reply]
Please don't use that `map` function as @folke mentioned. Just use `vim.keymap.set` for your mappings.

**Reply by dpetka2001** [reply]
Start a new `lua/config/keymaps.lua` file and set all your keymaps with `vim.keymap.set` or comment out all your mappings in your current file, so that there are no conflicting mappings.

**Reply by ConanEdogawa1** [reply]
I'll try

**Reply by dpetka2001** [reply]
Start with this in your new `lua/config/keymaps.lua`
```lua
local map = vim.keymap.set

map("n", "u", "<Up>")
map("n", "l", "<Undo>")
```
and then add whatever other mappings you would like.

**Reply by ConanEdogawa1** [reply]
Need to add `expr = true,noremap=true` ?

**Reply by folke** [reply]
For those mappings, you don't need expr I think, but honestly not sure.

**Reply by ConanEdogawa1** [reply]
Then I'll add it

**Reply by dpetka2001** [reply]
I think you don't need to add this. I tried without them and it worked for me.

**Reply by dpetka2001** [reply]
`expr = true`, as far as I understand it, is only needed when you want to return an expression being validated. This is not the case in your situation.

**Reply by ConanEdogawa1** [reply]
ok,thanks

**Reply by dpetka2001** [reply]
`:h vim.keymap.set()` has an example for Expr mappings
```lua
       -- Expr mappings
       vim.keymap.set('i', '<Tab>', function()
         return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"  -- Expression being validated
       end, { expr = true })

```
This is not the same as what you want to do.

**Reply by ConanEdogawa1** [reply]
Thank you!

**Reply by dpetka2001** [reply]
Just start fresh a new file like I told you in the above [post](https://github.com/LazyVim/LazyVim/discussions/287#discussioncomment-5009052) and then add whatever else you like. I think this way you will not be as confused.

**Reply by ConanEdogawa1** [reply]
nice!!!

**Reply by ConanEdogawa1** [reply]
I know why, as long as the editing is similar to lua, java, python, the u key will be invalid, and it will not be when editing ordinary text
